### PR TITLE
feat(workflow): Create partial index for AlertRule model

### PR DESCRIPTION
**Pull request by iProgramStuff**
We previously had a constraint requiring our alert rules to be unique by key (organization, name), but as we move forward with our new system we'll need to have alert rules that have the same name (as we archive alerts). However, we still want to enforce this constraint on active alerts so that active alerts will never have the same name.

Hence, a partial index that requires a unique key (organization, name, status) if status=0, but if status is something else (eg. 4, archived), then we don't enforce this.

Output of sqlmigrate:
```
--
-- Custom state/database change combination
--

                    ALTER TABLE sentry_alertrule DROP CONSTRAINT IF EXISTS sentry_alertrule_organization_id_name_12c48b37_uniq;
                    DROP INDEX CONCURRENTLY IF EXISTS sentry_alertrule_organization_id_name_12c48b37_uniq;
                    CREATE UNIQUE INDEX CONCURRENTLY sentry_alertrule_status_active
                    ON sentry_alertrule USING btree (organization_id, name, status)
                    WHERE status = 0;
                    ;
--
-- Alter field organization on alertrule
--
SET CONSTRAINTS "sentry_alertrule_organization_id_5eeaf3d8_fk_sentry_or" IMMEDIATE; ALTER TABLE "sentry_alertrule" DROP CONSTRAINT "sentry_alertrule_organization_id_5eeaf3d8_fk_sentry_or";
CREATE INDEX "sentry_alertrule_organization_id_5eeaf3d8" ON "sentry_alertrule" ("organization_id");
ALTER TABLE "sentry_alertrule" ADD CONSTRAINT "sentry_alertrule_organization_id_5eeaf3d8_fk_sentry_or" FOREIGN KEY ("organization_id") REFERENCES "sentry_organization" ("id") DEFERRABLE INITIALLY DEFERRED;
```

### CODE REVIEW

File: src/sentry/migrations/0061_alertrule_partial_index.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        ('sentry', '0060_add_file_eventattachment_index'),
+    ]
+
+    operations = [
+        migrations.SeparateDatabaseAndState(
+            database_operations=[
+                migrations.RunSQL("""
+                    ALTER TABLE sentry_alertrule DROP CONSTRAINT IF EXISTS sentry_alertrule_organization_id_name_12c48b37_uniq;
+                    DROP INDEX IF EXISTS sentry_alertrule_organization_id_name_12c48b37_uniq;
+                    CREATE UNIQUE INDEX sentry_alertrule_status_active
```
Comment (markstory) [inline src/sentry/migrations/0061_alertrule_partial_index.py:37]: Should this be created concurrently? This looks like a blocking index creation.

File: src/sentry/migrations/0061_alertrule_partial_index.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                    ON sentry_alertrule USING btree (organization_id, name, status)
+                    WHERE status = 0;
+                    """,
+                    reverse_sql="""
+                    DROP INDEX IF EXISTS sentry_alertrule_status_active;
+                    CREATE UNIQUE INDEX sentry_alertrule_organization_id_name_12c48b37_uniq
```
Comment (markstory) [inline src/sentry/migrations/0061_alertrule_partial_index.py:43]: Same here with the blocking index creation.

File: src/sentry/migrations/0061_alertrule_partial_index.py (hunk 1 above)
Comment (iProgramStuff) [reply src/sentry/migrations/0061_alertrule_partial_index.py:37]: Probably. Do you think I should also mark it as is_dangerous? I'm in doubt, so leaning towards yes.
`General advice is that if in doubt, mark your migration as `is_dangerous`.`

Comment (wedamija) [reply src/sentry/migrations/0061_alertrule_partial_index.py]: I don't think you need to mark this as `is_dangerous`, this table is fairly small and there's not a lot of risk.

File: src/sentry/migrations/0061_alertrule_partial_index.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    dependencies = [
+        ('sentry', '0060_add_file_eventattachment_index'),
+    ]
+
+    operations = [
+        migrations.SeparateDatabaseAndState(
+            database_operations=[
+                migrations.RunSQL("""
+                    ALTER TABLE sentry_alertrule DROP CONSTRAINT IF EXISTS sentry_alertrule_organization_id_name_12c48b37_uniq;
+                    DROP INDEX IF EXISTS sentry_alertrule_organization_id_name_12c48b37_uniq;
```
Comment (wedamija) [inline src/sentry/migrations/0061_alertrule_partial_index.py:36]: Let's drop this concurrently too

**Review by wedamija — COMMENTED** [review]
Can you remove `db_index=False` from the organization relation? Since we're only indexing on `(organization, name, status)` when status=0 we'll need that index in place.

File: src/sentry/migrations/0061_alertrule_partial_index.py (hunk 3 above)
Comment (iProgramStuff) [reply src/sentry/migrations/0061_alertrule_partial_index.py]: Done and done :-)

File: src/sentry/migrations/0062_add_alertrule_org_index.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    # want to create an index concurrently when adding one to an existing table.
+    atomic = True
+
+
+    dependencies = [
+        ('sentry', '0061_alertrule_partial_index'),
+    ]
+
+    operations = [
+        migrations.AlterField(
```
Comment (wedamija) [inline src/sentry/migrations/0062_add_alertrule_org_index.py:34]: I'd just add this as an operation to the original migration, you can add it before or after the `SeparateDatabaseAndState` one.

**Review by wedamija — COMMENTED** [review]
Looks good, but you missed a couple imports after merging the migrations. Once they're running, could you post the output of `sqlmigrate` just so we can confirm it's doing what we expect?

**Review by wedamija — APPROVED** [review]
Looks good. That index on org is not being created concurrently, but this table has 26 rows and is rarely accessed so I'm not concerned about table locks.

**Comment by iProgramStuff** [comment]
> Looks good. That index on org is not being created concurrently, but this table has 26 rows and is rarely accessed so I'm not concerned about table locks.

I noticed that. Is the only way to achieve that by manually writing the SQL again?

**Comment by wedamija** [comment]
> > Looks good. That index on org is not being created concurrently, but this table has 26 rows and is rarely accessed so I'm not concerned about table locks.
> 
> I noticed that. Is the only way to achieve that by manually writing the SQL again?

Either that or upgrade to 3.0 :D https://docs.djangoproject.com/en/3.0/ref/contrib/postgres/operations/#index-concurrent-operations

**Pull request merged**